hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java

```diff
@@ -410,61 +416,183 @@ public boolean rename(final Path src, final Path dst) throws IOException {
         listener);
     // rename under same folder;
-    if(makeQualified(parentFolder).equals(qualifiedDstPath)) {
+    if (makeQualified(parentFolder).equals(qualifiedDstPath)) {
       return tryGetFileStatus(qualifiedSrcPath, tracingContext) != null;
     }
 
-    FileStatus dstFileStatus = null;
     if (qualifiedSrcPath.equals(qualifiedDstPath)) {
       // rename to itself
       // - if it doesn't exist, return false
       // - if it is file, return true
       // - if it is dir, return false.
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
-      if (dstFileStatus == null) {
+      final AtomicBoolean isDstDirectory = new AtomicBoolean();
+      final AtomicBoolean isDstExists = new AtomicBoolean();
+
+      getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+          isDstExists);
+      if (!isDstExists.get()) {
         return false;
       }
-      return dstFileStatus.isDirectory() ? false : true;
+      return isDstDirectory.get() ? false : true;
     }
 
+    final AtomicBoolean isDstDirectory = new AtomicBoolean();
+    final AtomicBoolean isDstExists = new AtomicBoolean();
+
     // Non-HNS account need to check dst status on driver side.
-    if (!abfsStore.getIsNamespaceEnabled(tracingContext) && dstFileStatus == null) {
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
+    if (!abfsStore.getIsNamespaceEnabled(tracingContext)) {
+      getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+          isDstExists);
     }
 
     try {
       String sourceFileName = src.getName();
       Path adjustedDst = dst;
 
-      if (dstFileStatus != null) {
-        if (!dstFileStatus.isDirectory()) {
+      if (isDstExists.get()) {
+        if (!isDstDirectory.get()) {
           return qualifiedSrcPath.equals(qualifiedDstPath);
         }
         adjustedDst = new Path(dst, sourceFileName);
+        qualifiedDstPath = makeQualified(adjustedDst);
+        LOG.debug("Qualified dst path: {}", qualifiedDstPath);
+        /*
+         * For blob-endpoint with nonHNS account, check if the qualifiedDstPath
+         * exist in backend. If yes, HTTP_CONFLICT exception has to be thrown.
+         */
+        if (getAbfsStore().getAbfsConfiguration().getPrefixMode()
+            == PrefixMode.BLOB) {
+          isDstDirectory.set(false);
+          isDstExists.set(false);
+          getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+              isDstExists);
+          if (isDstExists.get()) {
+            //destination already there. Rename should not be overwriting.
+            LOG.info("Rename src: {} dst: {} failed as qualifiedDst already exists",
+                qualifiedSrcPath, qualifiedDstPath);
+            throw new AbfsRestOperationException(
+                HttpURLConnection.HTTP_CONFLICT,
+                AzureServiceErrorCode.PATH_ALREADY_EXISTS.getErrorCode(), null,
```

Comment:
Is the error message null in the case of DFS endpoint as well ? The error message should be same as DFS endpoint and not null.

Reply:
error code is same as DFS response.
error message doesn't matter, because it will be absorbed by https://github.com/saxenapranav/hadoop/blob/3af02586dfe553469dd040ca19b7a430336a2c3b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java#L529-L539,
https://github.com/saxenapranav/hadoop/blob/3af02586dfe553469dd040ca19b7a430336a2c3b/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java#L1546-L1548

The rename method will give false and not throw exception